### PR TITLE
Adding ConstrainedInput

### LINKED ISSUE #793: Constrained Input
The constrained input will be an opinionated text-input that takes a config / ruleset and generates the appropriate validation. This could be used for username / password etc inputs.

We should support common restrictions via a property and generate the appropriate validation patterns automatically. This should work along the lines of [password-sheriff](https://github.com/auth0/password-sheriff)

### PR DESCRIPTION
**Pull request by rorticus**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/framework/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**

Adding `ConstrainedInput` widget. This widget is just a regular `TextInput` with some predefined validation rules that will be used as base for the `PasswordInput`.

I've created a new `validation` middleware that defines these rules, thinking that we may want to use these validation rules in other widgets in the future.  It should work with any widget that has an `onValidate` property.

#### Docs
image

image

#### Contains Validation

Resolves #793

### CODE REVIEW

File: src/constrained-input/tests/unit/ConstrainedInput.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,105 @@
+import harness from '@dojo/framework/testing/harness';
+
+import ConstrainedInput from '../..';
+import { tsx, create } from '@dojo/framework/core/vdom';
+import TextInput from '../../../text-input';
+import validation from '../../../middleware/validation';
+
+const { registerSuite } = intern.getInterface('object');
+// const { assert } = intern.getPlugin('chai');
```
Comment (agubler) [inline src/constrained-input/tests/unit/ConstrainedInput.spec.tsx]: commented codes 😄

Comment (rorticus) [reply src/constrained-input/tests/unit/ConstrainedInput.spec.tsx]: GAH! Nothing to see here.

File: src/constrained-input/tests/unit/ConstrainedInput.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,105 @@
+import harness from '@dojo/framework/testing/harness';
+
+import ConstrainedInput from '../..';
+import { tsx, create } from '@dojo/framework/core/vdom';
+import TextInput from '../../../text-input';
+import validation from '../../../middleware/validation';
+
+const { registerSuite } = intern.getInterface('object');
```
Comment (agubler) [inline src/constrained-input/tests/unit/ConstrainedInput.spec.tsx]: Do we not use the `bdd` interfaces for widgets? I bet it's a mix unfortunately, I think we prefer the `bdd` template... we should certainly pick one and stick to it throughout the repo.

File: src/examples/src/widgets/constrained-input/Basic.tsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+					length: {
+						min: 2,
+						max: 8
+					}
+				}}
+				value={value1}
+				onValue={(value) => icache.set('value1', value)}
+				label="Length"
+			/>
+			<ConstrainedInput
```
Comment (agubler) [inline src/examples/src/widgets/constrained-input/Basic.tsx]: Should these be two different examples?

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* dojo.widgets: https://dojowidgets-2yid6g28y.now.sh

File: src/constrained-input/tests/unit/ConstrainedInput.spec.tsx (hunk 2 above)
Comment (rorticus) [reply src/constrained-input/tests/unit/ConstrainedInput.spec.tsx]: Soooo, yeah I prefer the bdd interface, but the other widgets I looked at do it this way. If we're ok using bdd i'm happy to switch it over...

File: src/middleware/validation.ts [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+const upperCaseRegEx = /[A-Z]/;
+const numbersRegEx = /\d/;
+
+export interface ValidationRules {
+	length?: {
+		min?: number;
+		max?: number;
+	};
+	contains?: {
```
Comment (agubler) [inline src/middleware/validation.ts]: If you specify `contains`, I think at least one of the options should be specified?

File: src/examples/src/widgets/constrained-input/Basic.tsx (hunk 3 above)
Comment (rorticus) [reply src/examples/src/widgets/constrained-input/Basic.tsx]: They certainly can be. I originally had a Basic example and a "Password" example that showed how you could use it as a password input. Problem was that they were so simple I wondered if they should be combined. I'm happy to separate them back out though!

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* dojo.widgets: https://dojowidgets-2yid6g28y.now.sh

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* dojo.widgets: https://dojowidgets-2yid6g28y.now.sh

File: src/middleware/validation.ts (hunk 4 above)
Comment (agubler) [reply src/middleware/validation.ts]: was thinking something like this: http://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgDIRAczACwLKgBKArgDYoDeAUMrcgLagBcyIx9ARtANw131wAHgH4WbTjyoBfKlVCRYiFOiy48QkuWTU6DUKNbsuUXroGCxRyTLnho8JMgDCAe3BxQAZwCCYdHE8wbT5aOD8IALBLCRMQ5GIABwToBACIA3FjUzpM6E8Mq1jdT2SEYDhSJxw4KEQFfOis6Vl5eyVnNzAPEE8AVSToJzTg3UTkqFTPCEbJXTD-QIKY7NpcqAbDZbiSiDKKqpq6vKWmm1bFR1d3LwA5Qs8RnPuZorp5iMWXlfiBibST2Z0HZ7SrVWoIeoA2JnOwXFBXLpeADKpXKoMOELyj1owLRB3B9S+cXekSh3zGKX+RN0aw2a14NggggSLigQXODhQADUKsAACZhYBuTQQB7UACQ5FUOAMKmw+CIZBQAB80Bh5epBCLeOKEJ1uhsEQbfAsgqqjbd7shzfrkaj9mCjutrR1rj1+uMhlMGbI9T0gnAWDzSPzBcKlQ8ALzY5BS+UsCgyGx+wLIDhB3kCsBCkAiqMxuO4BNxMzMZAARjiSd9blTCAzIazObzyGjOjohZwxd0ZiELErumrVBTQT5DdD2fD5Hz7done7Pdo5n7AAYADQl-hlgd0Icj5DTZDBifNiOtmN+xE9BN72tBGDjptT0Xn2fIS8GheLkmBftV5rDneyCYI+Ya5mebZxB+Xhfj2tJ-oOAH7l2R6ZmBLaQbo0HXjGxT2uiBJ5Ahu5IUBwCgZO4HTq+UG2jhb6jL8kyHjutC3v6yDAAAVhRp7UZhdDYZ4sGMeMzH9hui6rM8yAAMz-lIQA

Thanks ts playground for the massive url

File: src/examples/src/widgets/constrained-input/Basic.tsx (hunk 3 above)
Comment (rorticus) [reply src/examples/src/widgets/constrained-input/Basic.tsx]: OK, I separated the examples out! There is now a simple "Basic" example, and then a new "Password" example that shows you how to use it as a password input (even though soon we will have a PasswordInput widget.... ignore that part)

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* dojo.widgets: https://dojowidgets-onvszuqsb.now.sh

File: src/middleware/validation.ts (hunk 4 above)
Comment (rorticus) [reply src/middleware/validation.ts]: Done! Much nicer types ❤️

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* dojo.widgets: https://dojowidgets-gtw2c82oa.now.sh

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* dojo.widgets: https://dojowidgets-ocrkby2oo.now.sh

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* dojo.widgets: https://dojowidgets-ocrkby2oo.now.sh

File: src/examples/src/widgets/constrained-input/Password.tsx [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+					atLeast: 2,
+					uppercase: 1,
+					specialCharacters: 1,
+					numbers: 1
+				}
+			}}
+			value={value}
+			onValue={(value) => icache.set('value', value)}
+			label="Enter Password"
+			type="password"
```
Comment (tomdye) [inline src/examples/src/widgets/constrained-input/Password.tsx]: How about changing this to be a `Username` example? I think having the `password` type makes it had to reason with the text you're entering in terms of providing users with an example.

Comment (rorticus) [reply src/examples/src/widgets/constrained-input/Password.tsx]: Done!

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* dojo.widgets: https://dojowidgets-jped3d9pu.now.sh

File: src/constrained-input/index.tsx [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+	const validator = validation(rules);
+
+	const handleValidation = (valid?: boolean, message?: string) => {
+		icache.set('valid', { valid, message });
+
+		onValidate && onValidate(valid);
+	};
+
+	return (
+		<TextInput
```
Comment (tomdye) [inline src/constrained-input/index.tsx]: can you make this themeable too please and use the theme composition pattern we've used in the button variants and number / email input.

Comment (rorticus) [reply src/constrained-input/index.tsx]: Done!

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* dojo.widgets: https://dojowidgets-jush0kllv.now.sh

File: src/constrained-input/index.tsx [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+	onValidate?: (valid?: boolean) => void;
+}
+
+const factory = create({ icache, validation, theme }).properties<ConstrainedInputProperties>();
+export const ConstrainedInput = factory(function ConstrainedInput({
+	middleware: { icache, validation, theme },
+	properties
+}) {
+	const { rules, onValidate, ...props } = properties();
+	const valid = icache.get<{ valid?: boolean; message?: string }>('valid');
```
Comment (tomdye) [inline src/constrained-input/index.tsx]: could you use `createICacheMiddleware` instead and pass the generic there?

File: src/examples/src/widgets/constrained-input/Username.tsx [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+const factory = create({ icache });
+
+export default factory(function Basic({ middleware: { icache } }) {
+	const value = icache.getOrSet('value', '');
+
+	return (
+		<ConstrainedInput
+			rules={{
+				length: {
+					min: 4,
```
Comment (tomdye) [inline src/examples/src/widgets/constrained-input/Username.tsx]: i think this should still include special chars / uppercase etc to show the usage

Comment (rorticus) [reply src/examples/src/widgets/constrained-input/Username.tsx]: Maybe i'll rename it to something that isn't username related then, just something real generic? I don't think it makes sense for a username example to have to include special characters.

File: src/constrained-input/index.tsx (hunk 7 above)
Comment (rorticus) [reply src/constrained-input/index.tsx]: Done!

File: src/examples/src/widgets/constrained-input/Username.tsx (hunk 8 above)
Comment (rorticus) [reply src/examples/src/widgets/constrained-input/Username.tsx]: Ended up just adding those rules back 👍

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* dojo.widgets: https://dojowidgets-iiflj0t13.now.sh

**Pull request merged**